File: rq1/scripts/table_3.py
"""
reproduce_table3.py — CausalImpact on classified Reddit content.

Inputs (in rq1/data/):
  vpn_time_series_classified_hp.csv
  politics_time_series_classified_hp.csv

Output: rq1/data/table3_reproduced.csv

Usage:
  python rq1/reproduce_table3.py
"""
import os, warnings
os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np

if not hasattr(pd.DataFrame, "applymap"):
    pd.DataFrame.applymap = pd.DataFrame.map

from causalimpact import CausalImpact

DATA_DIR = os.path.join("rq1", "data")

MILESTONES = {
    "Oct2023": pd.Timestamp("2023-10-23"),
    "Mar2025": pd.Timestamp("2025-03-17"),
    "Jul2025": pd.Timestamp("2025-07-21"),
}

def run_ci(df_indexed, post_start):
    pre_start = df_indexed.index.min()
    pre_end = post_start - pd.Timedelta(days=7)
    post_end = post_start + pd.Timedelta(days=7 * 7)
    ci = CausalImpact(df_indexed, [pre_start, pre_end], [post_start, post_end], alpha=0.05)
    s = ci.summary()
    rel = p = None
    for line in s.split("\n"):
        if "Relative effect" in line and rel is None:
            for tok in line.split():
                try: rel = float(tok.strip("%()").replace(",", "")); break
                except: pass
        if "tail-area" in line:
            for tok in line.split():
                try:
                    v = float(tok)
                    if 0 <= v <= 1: p = v
                except: pass
    return rel, p

def sig(rel, p):
    if rel is None: return "ERROR"
    mark = "**" if p and p < 0.05 else ("*" if p and p < 0.10 else "")
    return f"{rel:+.1f}%{mark}"

if __name__ == "__main__":
    files = {
        "VPN": "vpn_time_series_classified_hp.csv",
        "Politics": "politics_time_series_classified_hp.csv",
    }

    rows = []
    for group, fname in files.items():
        path = os.path.join(DATA_DIR, fname)
        if not os.path.exists(path):
            print(f"SKIP {group}: {fname} not found"); continue
        df = pd.read_csv(path)
        df["week"] = pd.to_datetime(df["week"])
        df_i = df.set_index("week")[["total"]].astype(float)

        for ms_name, ms_date in MILESTONES.items():
            print(f"  {group} / {ms_name} ...", end=" ", flush=True)
            rel, p = run_ci(df_i, ms_date)
            print(sig(rel, p))
            rows.append({"group": group, "milestone": ms_name,
                         "relative_effect_pct": round(rel, 1) if rel else None,
                         "p_value": round(p, 4) if p else None})

    out = pd.DataFrame(rows)
    out.to_csv(os.path.join(DATA_DIR, "table3_reproduced.csv"), index=False)
    print(f"\n{out.to_string(index=False)}")
    print(f"\nSaved: {DATA_DIR}/table3_reproduced.csv")

File: rq1/scripts/table_3.R
library(CausalImpact)
library(zoo)

DATA_DIR <- "rq1/data"

MILESTONES <- list(
  Oct2023 = as.Date("2023-10-23"),
  Mar2025 = as.Date("2025-03-17"),
  Jul2025 = as.Date("2025-07-21")
)

snap_to_week <- function(target, weeks) {
  candidates <- weeks[weeks >= target]
  if (length(candidates) > 0) return(min(candidates))
  return(max(weeks[weeks < target]))
}

run_ci <- function(y, weeks, treatment_target) {
  td <- snap_to_week(treatment_target, weeks)
  pre_end <- td - 7; post_end <- td + 7 * 7
  idx <- weeks >= min(weeks) & weeks <= post_end
  s <- zoo(y[idx], order.by = weeks[idx])
  if (length(unique(y[idx])) <= 1) return(c(NA, NA))
  set.seed(20260827)
  tryCatch({
    ci <- CausalImpact(s, c(min(weeks), pre_end), c(td, post_end))
    avg <- ci$summary["Average", ]
    c(round(avg[["RelEffect"]] * 100, 1), round(avg[["p"]], 4))
  }, error = function(e) { cat("  ERR:", conditionMessage(e), "\n"); c(NA, NA) })
}

fmt <- function(r) {
  if (is.na(r[1])) return("NA")
  sig <- ifelse(!is.na(r[2]) & r[2] < 0.05, "**", ifelse(!is.na(r[2]) & r[2] < 0.10, "*", ""))
  sprintf("%+.1f%%%s", r[1], sig)
}

# ================================================================
# CONTENT — from table13_weekly_series.csv
# ================================================================
cat("================================================================\n")
cat("TABLE 3 — Content (posts + comments)\n")
cat("================================================================\n\n")

weekly <- read.csv(file.path(DATA_DIR, "table13_weekly_series.csv"), stringsAsFactors = FALSE)
weekly$week <- as.Date(weekly$week)

content_rows <- list(
  list(group = "VPN",      filter = "Raw"),
  list(group = "VPN",      filter = "Raw (UK)"),
  list(group = "VPN",      filter = "Classified"),
  list(group = "VPN",      filter = "Classified (UK)"),
  list(group = "Politics", filter = "Raw"),
  list(group = "Politics", filter = "Raw (UK)"),
  list(group = "Politics", filter = "Classified"),
  list(group = "Politics", filter = "Classified (UK)")
)

results <- data.frame()
for (cr in content_rows) {
  sub <- weekly[weekly$group == cr$group & weekly$filter == cr$filter, ]
  sub <- sub[order(sub$week), ]
  if (nrow(sub) == 0) { cat("SKIP:", cr$group, cr$filter, "\n"); next }
  for (ms in names(MILESTONES)) {
    cat(sprintf("  %s %s / %s ... ", cr$group, cr$filter, ms))
    r <- run_ci(sub$total, sub$week, MILESTONES[[ms]])
    cat(fmt(r), "\n")
    results <- rbind(results, data.frame(group = cr$group, filter = cr$filter,
      column = "Content", milestone = ms, relative_effect_pct = r[1], p_value = r[2]))
  }
}

# ================================================================
# USERS — from table12_user_weekly_series.csv (all filter levels)
# ================================================================
cat("\n================================================================\n")
cat("TABLE 3 — Users (new_users)\n")
cat("================================================================\n\n")

user_file <- file.path(DATA_DIR, "table12_user_weekly_series.csv")
if (!file.exists(user_file)) {
  cat("WARNING: table12_user_weekly_series.csv not found.\n")
  cat("Run rebuild_all_user_timeseries.py first.\n")
} else {
  user_weekly <- read.csv(user_file, stringsAsFactors = FALSE)
  user_weekly$week <- as.Date(user_weekly$week)

  user_rows <- list(
    list(group = "VPN",      filter = "Raw"),
    list(group = "VPN",      filter = "Raw (UK)"),
    list(group = "VPN",      filter = "Classified"),
    list(group = "VPN",      filter = "Classified (UK)"),
    list(group = "Politics", filter = "Raw"),
    list(group = "Politics", filter = "Raw (UK)"),
    list(group = "Politics", filter = "Classified"),
    list(group = "Politics", filter = "Classified (UK)")
  )

  for (ur in user_rows) {
    sub <- user_weekly[user_weekly$group == ur$group & user_weekly$filter == ur$filter, ]
    sub <- sub[order(sub$week), ]
    if (nrow(sub) == 0) { cat("SKIP:", ur$group, ur$filter, "(no data)\n"); next }
    for (ms in names(MILESTONES)) {
      cat(sprintf("  %s %s / %s ... ", ur$group, ur$filter, ms))
      r <- run_ci(sub$new_users, sub$week, MILESTONES[[ms]])
      cat(fmt(r), "\n")
      results <- rbind(results, data.frame(group = ur$group, filter = ur$filter,
        column = "Users", milestone = ms, relative_effect_pct = r[1], p_value = r[2]))
    }
  }
}

write.csv(results, file.path(DATA_DIR, "table3_reproduced.csv"), row.names = FALSE)
cat(sprintf("\nSaved: table3_reproduced.csv (%d rows)\n", nrow(results)))
print(results, row.names = FALSE)
